### DISCUSSION TITLE
[BP29] Favoriser les polices standards

### DISCUSSION THREAD

**Discussion by florinesueur** [cnumr/best-practices#227] [Bonnes pratiques] [answered]
Est-il possible de s'inspirer de la [BP15](https://github.com/cnumr/best-practices/blob/main/chapters/BP_015_fr.md) "Choisir les technologies les plus adaptées" pour établir un classement des solutions disponibles, classées de la plus à la moins performante en termes de green IT ?

Je pense : 
- polices standards
- polices d'Api (Google, Adobe, ...)
- polices customs (préciser les formats nécessaires)
- variables fonts

Par exemple dans l'alternative de la BP, nous utilisons la Montserra qui est [extrêmement utilisée ](https://fonts.google.com/analytics). Je me pose la question : est-il mieux de l'utiliser en locale comme l'exemple ou d'utiliser l'api Google sachant que les [Google Fonts sont mises en cache pendant 1 an sur les navigateurs](https://developers.google.com/fonts/faq#what_does_using_the_google_fonts_api_mean_for_the_privacy_of_my_users) ?

Et peut-être aussi ajouter un `font-display: fallback` ou est ce que c'est une BP indépendante ? 
Références : [MDN](https://developer.mozilla.org/en-US/docs/Web/CSS/@font-face/font-display) et [Google](https://developers.google.com/web/updates/2016/02/font-display)

**Comment by tbroyer** [comment]
Sur la mise en cache, cf. https://github.com/cnumr/best-practices/discussions/83#discussioncomment-1531227
Du coup, puisqu'en plus ce sera en général plus performant (https://web.dev/font-best-practices/#using-self-hosted-fonts, https://wicki.io/posts/2020-11-goodbye-google-fonts/, https://www.peakhour.io/blog/cache-partitioning-firefox-chrome/), je pense qu'il vaut mieux l'héberger soi-même (sans compter que Google Fonts n'est _a priori_ pas RGPD-compliant: https://github.com/google/fonts/issues/1495).

Je ne connais pas assez les variable fonts, mais a priori (https://web.dev/variable-fonts/#performance-gains) il faudrait les privilégier (dès qu'on n'utilise pas les polices standards)
…sauf si bien sûr le fichier inclut des variations qui ne sont pas utilisées, il faudra du coup penser à "optimiser la font" (ça vaut dans tous les cas d'ailleurs, variable font ou non; pour ne pas inclure les glyphes de tous les scripts possibles sur un site essentiellement en alphabet latin par exemple, et ça peut faire l'objet d'une BP spécifique du coup, comme pour l"optimisation des images)

Concernant `font-display: fallback`, ça ne me semble pas lié à l'écoconception :thinking: (dans le sens où la font sera téléchargée de toutes façons)

**Reply by florinesueur** [reply]
Merci beaucoup pour les ressources donc il vaut mieux héberger soi-même.

Le font-display va plutôt dans le sens faire gagner du temps à l'utilisateur. "Par défaut, les navigateurs basés sur Chromium et Firefox bloquent le rendu du texte jusqu'à 3 secondes si la police Web associée n'a pas été chargée ; Safari bloquera le rendu du texte indéfiniment." Mais c'est du détail.

J'ai testé les utilisations d'une police au hasard : la Oswald.
| Source  | Fontes         | Poids | Temps |
| :--------------- |:---------------:|:------------:| -----:|
| Api Google  |  Regular        |  16.0kb | 32ms |
| Api Google  | Regular + Bold   | 24.1kb     |   62ms |
| Api Google  | Variable          | 24.1kb | 33ms|
|***|***|***|***|
| Local en woff2  |   Regular        |  37.6kb| 4ms |
| Local en woff2  | Regular + Bold   |     76.9kb     |   15ms |
| Local en woff2 | Variable          |    66kb | 6ms|
|***|***|***|***|
| Local en woff2 |  Latin Regular        |  16.0kb | 4ms |
| Local en woff2 |Latin Regular + Latin Bold   |     32.5kb     |   14ms |

Il serait possible de faire la version Latin variable via FontForge ou Glyphs.

Article explicatif pour convertir la police variable de TFF en woff2 : [https://henry.codes/writing/how-to-convert-variable-ttf-font-files-to-woff2/](https://henry.codes/writing/how-to-convert-variable-ttf-font-files-to-woff2/)
Le support des variables fonts : [https://caniuse.com/variable-fonts](https://caniuse.com/variable-fonts)

**Comment by ldesgrange** [comment]
On peut ajouter les polices systèmes à la liste (cf. https://systemfontstack.com/)

**Comment by Fred-artwai** [comment]
Limiter les polices aux polices systèmes c'est pour moi un bel exemple de décroissance web ;-)
Je pense comme @florinesueur on doit pouvoir utiliser une police envoyé par le site web. Alors pas 36 on n'est d'accord... et plutôt autohébergé. 
Je recommanderais les polices variables dans lesquels on peut combiner variantes de police et glyphes d'icones. Donc moins de fichier et souvent moins de poids.

Accessoirement, on peut aussi recommander d'utiliser le graissage de font en CSS plutôt que d'utiliser une police dédiée au "gras" : https://www.artwai.com/faux-gras-alleger-font-web/

NB : attention, ne pas recompresser les woff2 cotés serveur. Trop souvent je vois ces fichiers gzippé alors qu'ils sont déjà compressé par nature.

**Reply by arnolem** [reply]
Personnellement, je ne suis pas pour un retour systématique aux polices systèmes. Ca serait la même chose que de dire d'utiliser la mise en page par défaut des balises HTML (Les H1 sont déjà plus gros que les H2, pas besoin d'ajouter du CSS alors ?) 😁.

Toutefois, ça reste quand même une bonne pratique à proposer pour ceux qui le souhaite.
J'avais fait un test sur ce site et sincèrement, la police système n'est pas si "sale" : https://siteweb.eco/

Pour moi, il faut garder cela comme une proposition, une version extrême de l'optimisation des fonts, sans pour autant dire que d'utiliser des webFont serait une mauvaise pratique.

**Reply by ldesgrange** [reply]
Je disais juste d'ajouter ça à la liste des possibilités hein, pas qu'il fallait faire uniquement ça.

Ceci étant dit, pour une PWA qui veut avoir un look natif, utiliser les fonts systèmes me semble être le meilleur choix.

De même, pour tout ceux qui n'ont pas de besoin spécifique pour leurs polices, ça peut être un choix intéressant, les OS essayant en général d'être utilisables facilement et de ne pas être moche, les polices systèmes passent plutôt bien (et couvrent tous les caractères). D'ailleurs github utilise ça.

**Reply by tbroyer** [reply]
Booking ou Trainline utilisent aussi les polices système (pas partout, mais au moins pour la majorité du texte).

**Comment by ACTLEM** [comment, marked as answer]
Suite aux discussions, il serait intéressant de mentionner les différents "types" de polices, énoncer les polices systèmes, l'auto hébergement... J'ai donc créé l'issue:  https://github.com/cnumr/best-practices/issues/271
